Add document with the vision of the buildfarmer payload process (#22)
osrf/buildfarm-tools: vision.md

# Vision of the buildfarmer payload
Subset of tasks defined to increase the quality of the ROS/Gazebo code by actively tracking and categorizing problems shipped with new code in a timely manner, making it easier for the development teams to:
* Fix most errors when they are introduced.
* Decide if an error that pops up in CI is a flaky test already existing in the codebase.
* Identify how often an error occurs and decide how to proceed.

## Internal process and rationale behind it

We already have a priority document explaining the order of relevancy of the possible errors, this is a new set of standards to define what we expect of the internal buildfarmer process:

1. Auto-log any error appearing in the nightlies, even when there is no one around. This could serve future investigations.
2. Any person can become a Buildfarmer Payload in case of necessity, this implies:
   * Basic duties should be easy to automate, even scripted.

>    * Every Buildfarmer Payload must log/communicate internally any actions completed in the name of the role, so other folks can make progress on top of previous actions.

Reviewer: Same here it's odd saying payload when referencing a  person. Either it should be a buildfarmer or perform buildfarmer payload

   * There will be a public log with investigations related to the Buildfarmer Payload duties.
   * Actions related to Buildfarmer Payload will be differentiated from actions related to the infra team, it is not resposability of the role to investigate infra issues, instead those will be escalated.
3. Core Team and community developers should be able to check if an existent error in his build it’s also happening in the nightlies CI.

> 4. Flaky tests will be categorized based on how often they appear

Reviewer 2: Same comment I did on the document:

"How often they appear" in which timestamp? 

I think it is important to define in which time frame we're talking about (2 months, 1 month, 2 weeks), or in how many builds do we want to measure the percentage.

E.g., for a 1-month timeframe of a ROS nightly job: 5% is 1.55 builds, 15% is 4.65 builds.

In Gazebo, most of the jobs run each four days. We would need to calculate the flakiness in almost 3 months to maybe get a 5% flaky ratio.

Also, our current workflow is priotirizing the "Number of failed builds" over the "Flaky ratio".

Author: Been thinking a bit about this, here's the rationale I think we should follow in terms of the number of builds. I'm thinking in terms of 2 weeks of ROS 2 nightlies, which is around 14 builds.

3 errors in 14 builds =  21%
2 errors in 14 builds = 14%
1 error in 14 build = 7%

probably 20% is a good mark to start pushing actively for issues, and disabling them as needed. Anything below 7% should be ignored in terms of investigations, and anything in the middle should be help wanted to the developer teams.

> In Gazebo, most of the jobs run each four days. We would need to calculate the flakiness in almost 3 months to maybe get a 5% flaky ratio.

Hopefully we can improve that in the future, in the meantime the rationale to categorize the flakiness shouldn't vary. When calculating flakiness, we will be taking into account last month test failures and I added an extra section about prioritizing consecutive failures. 

The discussion about how to categorize the flakiness of an error will not be used to categorize new errors, as we simply won't have enough info to consider them flaky.

   * **~5%** Will not be reported.
   * **6% - 15%** Will be reported and announced in the weekly meetings. This category will be considered as known issues, as we won’t aim to disable those tests.
   * **15% - 100%** Will be reported, investigated internally, and eventually assigned someone from the development teams. Flaky tests falling in this category will be expected to be disabled 3 weeks after being assigned a developer to take a look or after being reported.
5. The buildfarmer team will keep track of all the tests disabled internally and generate a report monthly. That way any developer can track those and aim to fix the underlying issues.
6. As a new error gets investigated, make sure to close the reported issues/investigations ONLY after the issue is addressed in all combinations of job, platform, distribution where it was reported. Group errors in one a single issue if there's evidence they share an underlying reason (appeared after the same date, they occur in the same jobs, etc).
7. Reports to generate monthly:

>    * Buildfarm average greenness per distro / platform.

Author: Script TBD

Author: We can list the number of disabled tests as part of this greenness

>    * Disabled tests.

Author: @Crola1702 will take a look as to have much overhead will generate this to the workflow of the buildfarmer payload.

I imagine we can set up an automation to read from a column in our buildfarmer issues dashboards to read from there.

Author: Not only disabled tests for the month, ideally this should have a monthly summary of new items in our board: https://github.com/orgs/osrf/projects/23/views/1

And the process will force us to keep that updated with the investigations, tracking assignees, etc

>    * Average time job per OS.

Author: Script TBD, JJ to take a look with greenness script

>    * Instance rotation reports.

Author: We have a script to do this already, although might be a bit outdated. @claraberendsen To add a link here, no need to update the report tool for now.

Reviewer: Here is the link to the script that generates them, https://github.com/osrf/infrastructure-private/blob/main/infra/reports/autoscaling_metrics_report.py and the reports generated are on the infra drive under infra reports.

>    * Jenkins metrics reports.

Author: TBD

Author: We don't have any tooling in place to observe the state of the agents in jenkins without a significant engineering effort, we will postpone this for now.

# Buildfarmer Payload tasks process vision
Subset of tasks defined to increase the quality of the ROS/Gazebo code by actively tracking and categorizing problems shipped with new code in a timely manner, making it easier for the development teams to:
* Fix most errors when they are introduced.
* Decide if an error that pops up in CI is a flaky test already existing in the codebase.
* Identify how often an error occurs and decide how to proceed.

## Internal process and rationale behind it

We already have a priority document explaining the order of relevancy of the possible errors, this is a new set of standards to define what we expect of the internal buildfarmer process:

1. Auto-log any error appearing in the nightlies, even when there is no one around. This could serve future investigations.

> 2. Any person can be assigned to Buildfarmer Payload in case of necessity, this implies:

Reviewer 3: I haven't read the rest of the documents in here, but the "Buildfarmer Payload" role isn't defined prior to its use here.  I think we either need to define what it is first, or link to a document that does so.

In particular, it is unclear to me where the person assigned to the Buildfarm Payload role will come from.

Author: > the "Buildfarmer Payload" role isn't defined prior to its use here

The first paragraph is supposed to explain what's the Buildfarm Payload role, I'll add a bit more to make it explicit.

Author: > it is unclear to me where the person assigned to the Buildfarm Payload role will come from

This was deliberately loose, as of today I could state that this role will be assigned to someone from the infra team. 

However, our vision for the role is having all the tools available publicly and make it possible for anyone from the community to run take this role (if someone ever wanted to). I'll add a paragraph to track this.

   * Basic duties should be easy to automate, even scripted.
   * Everyone assigned to Buildfarmer Payload must log/communicate internally any actions completed in the name of the role, so other folks can make progress on top of previous actions.
   * There will be a public log with investigations related to the Buildfarmer Payload duties.

>    * Actions related to Buildfarmer Payload will be differentiated from actions related to the infra team, it is not resposability of the role to investigate infra issues, instead those will be escalated.

Reviewer 3: ```suggestion
   * Actions related to Buildfarmer Payload will be differentiated from actions related to the infra team, it is not the responsibility of this role to investigate infra issues, instead those will be escalated.
```

> 3. Core Team and community developers should be able to check if an existent error in his build it’s also happening in the nightlies CI.

Reviewer 4: ```suggestion
3. Core Team and community developers should be able to check if an existent error in their build is also happening in the nightlies CI.
```

> 3. Core Team and community developers should be able to check if an existent error in his build it’s also happening in the nightlies CI.

Reviewer 3: ```suggestion
3. Core Team and community developers should be able to check if an error in a local build is also happening in the nightly CI.
```

> 4. Consecutive test failures priorities policy: Test errors occuring consecutively will be prioritized in terms of order of investigation, in the case of Gazebo jobs, the failing build will be retriggered to confirm if the error occurs a third time in a row. Errors happening three times in a row will be reported immediatly to the development teams in case the buildfarmer investigation wasn't enough to solve it. Errors that have appeared at least three times every 2 weeks, will also be prioritized. Buildfarmer Payload scripts will prioritize accordingly new investigations based on this.

Reviewer 4: ```suggestion
4. Consecutive test failures priorities policy: Test errors occuring consecutively will be prioritized in terms of order of investigation. In the case of Gazebo jobs, the failing build will be retriggered to confirm if the error occurs a third time in a row. Errors happening three times in a row will be reported immediately to the development teams in case the buildfarmer investigation wasn't enough to solve it. Errors that have appeared at least three times in a 2 week window, will also be prioritized. Buildfarmer Payload scripts will prioritize accordingly new investigations based on this.
```

Not sure if my last edit conveys what you meant.

Author: Yes! thank you

5. Flaky tests will be categorized based on how often they appear in a given month in a unstable builds.
   * **~7% or errors that happened only once** Will not be reported.
   * **7% - 20%** Will be reported and announced in the weekly meetings. This category will be considered as known issues, as we won’t aim to disable those tests.

>    * **20% - 100%** Will be reported, investigated internally, and eventually assigned someone from the development teams. Flaky tests falling in this category will be expected to be disabled 3 weeks after being assigned a developer to take a look or after being reported.

Reviewer 3: I actually really like having numbers attached to this.  It makes it very clear, thanks.

6. The buildfarmer team will keep track of all the tests disabled internally and generate a report monthly. That way any developer can track those and aim to fix the underlying issues.
7. As a new error gets investigated, make sure to close the reported issues/investigations ONLY after the issue is addressed in all combinations of job, platform, distribution where it was reported. Group errors in one a single issue if there's evidence they share an underlying reason (appeared after the same date, they occur in the same jobs, etc).
8. Reports to generate monthly:
   * Buildfarm average greenness per distro / platform (TBD)

>    * Disabled tests: Those can be obtained using the jenkins API, will be saved to our internal database.

Reviewer 4: Can you verify that it is indeed possible to obtain this using the Jenkins API? For example, https://build.osrfoundation.org/view/gz-harmonic/job/gz_sim-8-win/24/testReport/ shows only 3 tests skipped, which I assume is what the API would tell you, but we have a lot more tests disabled on windows with the `DISABLED_` gtest prefix (see https://google.github.io/googletest/advanced.html#temporarily-disabling-tests). The ones that show up as skipped actually run but immediately call `GTEST_SKIP`.

Reviewer 5: I don't think that Jenkins can show disabled tests (versus skipped test) since the disabled tests produce no output in the xunit/junit xml which is what Jenkins uses to assemble its test reporting. We would have to use a different tool or see if we can tweak ctest/gtest to emit a SKIP result for disabled tests.

Author: Yes, Steven! is right, I will modify this text to reflect that we will track in the database skipped tests as tracked by Jenkins.

We don't have a way to track disabled tests by now automatically, ideally the dev teams will let us know whenever tests are being disabled so we can save them in our database and included it as part of the monthly report.
